design-documents/decrypt-api-token.md

```diff
@@ -0,0 +1,255 @@
+     * @throws TokenAuthenticationException If token can be read but cannot be authenticated.
+     */
+    public function decode(string $token, EncryptionDataInterface $encryption): array;
+}
+```
+ 
+There are 2 ways we can go about implementing this:
+ 
+#### Own implementation using EncyptorInterface
+Authentication data will be serialized into json, signed using hash with secret key provided by KeyStorageInterface,
+serialized data and hash paired and obfuscated.
+ 
+Magento's _Encryptor_ can be updated to allow using custom key or it's instance can be created with key provided to the
+constructor.
+ 
+Example in pseudo-code:
+```php
+$userData = ['user_data' => ['user_id' => 1, 'user_type' => 3], 'issued' => time(), 'expires' => time() + (60*60)];
+$encodedData = json_encode($userData);
+$signature = $encryptor->hash($encodedData);
+
+return $token = base64_encode($encodedData) .'.' .base64_encode($signature);
+```
+ 
+When decoding tokens signatures will be recreated and compared to provided ones to validate tokens' origin.
+ 
+__Pros__:
+* Easy to implement
+* Uses EncryptorInterface which guarantees up-to-date secure encryption
+* No new project dependencies introduced
+ 
+__Cons__:
+* Custom solution
+* May suffer from problems that an established standard solutions would have already solved
+ 
+#### Using JWT
+JWT is an established standard for authentication tokens that contain user data and are signed as origin verification.
+A number of [PHP libraries](https://jwt.io) existing implementing it.
+ 
+Child interfaces of `TokenEncoderInterface` will be created - `JWSTokenEncoderInterface` and `JWETokenEncoderInterface`
+that will contain JWT specific
+methods in addition to the default one that will allow, for instance, pick algorithms, use public and private claims etc.
+ 
+__Pros__:
+* Established standard
+* Existing libraries
+* Wide variety of supported encryption algorithms
+* Security concerns resolved by standard and libraries themselves
+* Easy to use for integrations
+ 
+__Cons__:
+* Will have to add new project dependency for a library and it's dependencies
+* Some libraries have [known security vulnerabilities](https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/)
+* Classes responsible for token creation will have to be updated alongside `EncryptorInterface` in order to utilize the
+  most secure and up-to-date algorithms
+ 
+[JWT Frameworks](https://github.com/web-token/jwt-framework) looks to be promising candidate to use for Magento implementation.
+ 
+##### API for JWT token encoder
+```php
+interface JwsTokenEncoderInterface extends TokenEncoderInterface
+{
+    /**
+     * JWS specific decoding method that returns more info regarding the token.
+     *
+     * @return JwsToken JWS DTO containing headers, payload and resolved token data.
+     */
+    public function decodeJws(string $token, JwsEncryptionDataInterface $encryption): JwsToken;
```

Comment:
It looks strange that encoder interface has method to decode.

Reply:
what is the problem with that? Makes sense to me that a class that can encode a value can decode it as well

Reply:
Yes, if this class is supposed to work like management but in your case, the interface, according to its name, has a single responsibility - encode data and the additional method `decodeJws` violates it.

Reply:
just updated couple of class names, please take a look @joni-jones